File: Tracks/typescript/queen-attack/queen-attack.ts
type Position = [number, number];
interface QueenPositions {
  black: Position;
  white: Position;
}

const Equals = (vector1: Position, vector2: Position): boolean =>
  vector1[0] === vector2[0] && vector1[1] === vector2[1];

export default class QueenAttack {
  public white: Position;
  public black: Position;

  constructor(positions: QueenPositions) {
    if (Equals(positions.white, positions.black)) {
      throw new Error("Queens cannot share the same space");
    }
    this.white = positions.white;
    this.black = positions.black;
  }

  toString = (): string =>
    [...Array(8)]
      .map((_, row) =>
        [...Array(8)]
          .map((_, column) =>
            Equals(this.white, [row, column])
              ? "W"
              : Equals(this.black, [row, column])
              ? "B"
              : "_"
          )
          .join(" ")
      )
      .join("\n") + "\n";

  canAttack(): boolean {
    const distance = [
      this.white[0] - this.black[0],
      this.white[1] - this.black[1],
    ];
    const absDistance = distance.map((coord) => Math.abs(coord));
    const maxDistance = Math.max(...absDistance);
    const normalizedAbs = absDistance.map((coord) => coord / maxDistance);
    return [
      [0, 1],
      [1, 0],
      [1, 1],
    ].some((vector) => Equals(vector as Position, normalizedAbs as Position));
  }
}

File: Tracks/csharp/queen-attack/QueenAttack.cs
using System;
using System.Collections.Generic;
using System.Numerics;

public class Queen
{
    public Queen(int row, int column)
    {
        if (row < 0 || column < 0 || row > 7 || column > 7) throw new ArgumentOutOfRangeException("Invalid positions");
        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class QueenAttack
{
    public static bool CanAttack(Queen white, Queen black)
    {
        var normalizedAbsolute = Vector2.Abs(Vector2.Normalize(new Vector2(Math.Abs(white.Row - black.Row), Math.Abs(white.Column - black.Column))));
        return new List<Vector2> { Vector2.UnitX, Vector2.UnitY, Vector2.Normalize(Vector2.One) }.Contains(normalizedAbsolute);
    }

    public static Queen Create(int row, int column) => new Queen(row, column);
}